### ISSUE TITLE
Cannot build gRPC/protobuf with WindRiver Linux Cross Toolchain on Ubuntu LTS

### ISSUE THREAD

**Issue by nmartelwork** [grpc/grpc#16052]
### What version of gRPC and what language are you using?
Branch v1.12.x 
 
### What operating system (Linux, Windows, …) and version?
 Development Platform is Ubuntu LTS 16.04
Target Platform is WindRiver Linux LTS 17 (i.e. Yocto 2.4.3)
BSP and target hardware is Intel x86_64

### What runtime / compiler are you using (e.g. python version or version of gcc)
 x86_64-wrs-linux-gcc version 7.3.0
 
### What did you do?
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.

Step 1) Built gRPC, third party elements natively.  Built my app and tested on Ubuntu -- worked well
Step 2) Installed cross toolchain generated in bitbake of the WindRiver/Yocto project
Step 3) Performed the following steps to get a cross-platform build of gRPC/protobuf that would enable me to build my app for the targeted OS/environment:

# Protobuf installation [ DEVELOPMENT HOST ]
cd ~/grpc/third-party/protobuf
./autogen.sh
./configure
# Consider disabling shared libraries to force static linking, especially for cross-toolchain builds
./configure --disable-shared

make && make check && sudo make install && sudo ldconfig

# Grpc installation [ DEVELOPMENT HOST ]
cd ~/grpc
make && sudo make install && sudo ldconfig

# Grpc compile for CROSS PLATFORM TARGET (Clean build)
cd ~/grpc
make clean
make plugins

# exort path to cross utilities and cross tools (gRPC Makefile does not pre-pend sysroot)
export PATH=/opt/windriver/wrlinux/10.17.41/sysroots/x86_64-wrlinuxsdk-linux/usr/bin:/opt/windriver/wrlinux/10.17.41/sysroots/x86_64-wrlinuxsdk-linux/usr/bin/x86_64-wrs-linux:$PATH

export GRPC_CROSS_COMPILE=true
export GRPC_CROSS_LDOPTS="-L/opt/windriver/wrlinux/10.17.41/sysroots/corei7-64-wrs-linux/usr/lib64"
export GRPC_CROSS_AROPTS="cr --target=elf32-little"

make HAS_PKG_CONFIG=false \
    CPPFLAGS="-I/home/nmartel/grpc/include/grpc++ -I/opt/windriver/wrlinux/10.17.41/sysroots/corei7-64-wrs-linux/usr/include" \
    CC=x86_64-wrs-linux-gcc \
    CXX=x86_64-wrs-linux-g++ \
    RANLIB=x86_64-wrs-linux-ranlib \
    LD=x86_64-wrs-linux-ld \
    LDXX=x86_64-wrs-linux-g++ \
    AR=x86_64-wrs-linux-ar \
    STRIP=x86_64-wrs-linux-strip \
    PROTOBUF_CONFIG_OPTS="--host=x86_64-wrs-linux-gnueabi --with protoc=/usr/local/bin/protoc" static

### What did you expect to see?
 Build gRPC and third party libraries that I could install into my cross toolchain to enable me to build/link my app to the targeted environment.
 
### What did you see instead?
The build failed immediately with the following output.  Seems that the cross toolchain environment and/or setting of variables prevents the gRPC build from finding its own header files.

nmartel@ubuntu:~/grpc$ make HAS_PKG_CONFIG=false \
>     CPPFLAGS="-I/home/nmartel/grpc/include/grpc++ -I/opt/windriver/wrlinux/10.17.41/sysroots/corei7-64-wrs-linux/usr/include" \
>     CC=x86_64-wrs-linux-gcc \
>     CXX=x86_64-wrs-linux-g++ \
>     RANLIB=x86_64-wrs-linux-ranlib \
>     LD=x86_64-wrs-linux-ld \
>     LDXX=x86_64-wrs-linux-g++ \
>     AR=x86_64-wrs-linux-ar \
>     STRIP=x86_64-wrs-linux-strip \
>     PROTOBUF_CONFIG_OPTS="--host=x86_64-wrs-linux-gnueabi --with protoc=/usr/local/bin/protoc" static VERBOSE=1
[MAKE]    Generating cache.mk
make: Circular /home/nmartel/grpc/libs/opt/libares.a <- /home/nmartel/grpc/libs/opt/libz.a dependency dropped.
make: Circular /home/nmartel/grpc/libs/opt/libares.a <- /home/nmartel/grpc/libs/opt/libaddress_sorting.a dependency dropped.
make: Circular /home/nmartel/grpc/libs/opt/libz.a <- /home/nmartel/grpc/libs/opt/libaddress_sorting.a dependency dropped.
[CXX]     Compiling src/core/lib/gpr/alloc.cc
src/core/lib/gpr/alloc.cc:19:10: fatal error: grpc/support/port_platform.h: No such file or directory
 #include <grpc/support/port_platform.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Makefile:2817: recipe for target '/home/nmartel/grpc/objs/opt/src/core/lib/gpr/alloc.o' failed
make: *** [/home/nmartel/grpc/objs/opt/src/core/lib/gpr/alloc.o] Error 1

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).
All you would need is my cross toolchain and a stock Ubuntu machine with the pre-reqs installed. 
I can provide the cross toolchain self-installer upon request.

**Comment by nmartelwork** [comment]
Interesting... I tried v1.13.1 branch after seeing a handful of reports indicating cross platform build issues.  This worked a tad better in that the first project ZLIB built, but then died in the alloc.cc compilation due to the same error as above.  The <grpc/support/port_platform.h> was not found.  What is this build looking for? I assume there is something in the environment that identifies what the target build is looking for and that maybe a path is altered based upon that target?
Here is the snip from the failure:
nmartel@ubuntu:~/grpc$ make HAS_PKG_CONFIG=false \
>     CPPFLAGS="-I/home/nmartel/grpc/include/grpc++ -I/opt/windriver/wrlinux/10.17.41/sysroots/corei7-64-wrs-linux/usr/include" \
>     CC=x86_64-wrs-linux-gcc \
>     CXX=x86_64-wrs-linux-g++ \
>     RANLIB=x86_64-wrs-linux-ranlib \
>     LD=x86_64-wrs-linux-ld \
>     LDXX=x86_64-wrs-linux-g++ \
>     AR=x86_64-wrs-linux-ar \
>     STRIP=x86_64-wrs-linux-strip \
>     PROTOBUF_CONFIG_OPTS="--host=x86_64-wrs-linux-gnueabi --with protoc=/usr/local/bin/protoc" static
[MAKE]    Generating /home/nmartel/grpc/libs/opt/pkgconfig/grpc.pc
[MAKE]    Generating /home/nmartel/grpc/libs/opt/pkgconfig/grpc_unsecure.pc
[MAKE]    Generating cache.mk
[C]       Compiling third_party/zlib/adler32.c
[C]       Compiling third_party/zlib/compress.c
[C]       Compiling third_party/zlib/crc32.c
[C]       Compiling third_party/zlib/deflate.c
[C]       Compiling third_party/zlib/gzclose.c
[C]       Compiling third_party/zlib/gzlib.c
[C]       Compiling third_party/zlib/gzread.c
[C]       Compiling third_party/zlib/gzwrite.c
[C]       Compiling third_party/zlib/infback.c
[C]       Compiling third_party/zlib/inffast.c
[C]       Compiling third_party/zlib/inflate.c
[C]       Compiling third_party/zlib/inftrees.c
[C]       Compiling third_party/zlib/trees.c
[C]       Compiling third_party/zlib/uncompr.c
[C]       Compiling third_party/zlib/zutil.c
[AR]      Creating /home/nmartel/grpc/libs/opt/libz.a
[CXX]     Compiling src/core/lib/gpr/alloc.cc
src/core/lib/gpr/alloc.cc:19:10: fatal error: grpc/support/port_platform.h: No such file or directory
 #include <grpc/support/port_platform.h>
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Makefile:2836: recipe for target '/home/nmartel/grpc/objs/opt/src/core/lib/gpr/alloc.o' failed
make: *** [/home/nmartel/grpc/objs/opt/src/core/lib/gpr/alloc.o] Error 1
nmartel@ubuntu:~/grpc$

**Comment by nmartelwork** [comment]
I've come back around to this task... it is not blocking my development.
I checked out what seems to be the latest stable branch (v1.14.x).
I checked out a fresh copy of gRPC with the third party tools.  Then, descended into protobuf.
Doing the following:
> make clean
> ./configure --disable-shared --with-protoc=protoc
> source /opt/windriver/wrlinux/10.17.41/environment-setup-core2-32-wrs-linux
sets me up with an appropriately configured environment for a cross build.
> make && make check 
these now finish cleanly.  So progress has been made in the build area (makefiles?).
I am now digging to see how to get the appropriate path for the make install.
Before building gRPC, I need the .la files installed into the toolchain that gRPC will be linking with.
this path is:  /opt/windriver/wrlinux/10.17.41/sysroots/i686-wrlinuxsdk-linux
Within this path, there is a minimal root file system with all the headers and libraries (as well as tools) needed to build code for the target.  (Note this would apply to ARM or any other cross target).

Any idea on how to specify the prefix so that install places the headers and .la files into /opt/windriver...
??
Thanks
